components/templates/src/global_fns/load_data.rs

```diff
@@ -0,0 +1,346 @@
+extern crate toml;
+extern crate serde_json;
+
+use utils::fs::{read_file, is_path_in_directory, get_file_time};
+
+use crypto_hash::{Algorithm, hex_digest};
+use chrono::{DateTime, Utc};
+use reqwest::{Client, header};
+use url::Url;
+
+use std::path::PathBuf;
+use std::sync::{Arc, Mutex};
+
+
+use csv::Reader;
+use std::collections::HashMap;
+use tera::{GlobalFn, Value, from_value, to_value, Result, Map};
+use std::ops::BitXor;
+
+static GET_DATA_ARGUMENT_ERROR_MESSAGE: &str = "`load_data`: requires EITHER a `path` or `url` argument";
+
+enum ProvidedArgument {
+    URL(Url),
+    PATH(PathBuf)
+}
+
+
+fn get_cache_key(provided_argument: &ProvidedArgument, format: &String) -> String {
+    let content_based_data = match provided_argument {
+        ProvidedArgument::URL(url) => url.clone().into_string(),
+        ProvidedArgument::PATH(path) => {
+            let file_time = get_file_time(&path).expect("get file time");
+            let file_datetime: DateTime<Utc> = DateTime::from(file_time);
+            format!("{}{}", file_datetime.timestamp_millis().to_string(), path.display())
+        }
+    };
+    return hex_digest(Algorithm::MD5, format!("{}{}", format, content_based_data).as_bytes());
+}
+
+
+fn get_data_from_args(content_path: &PathBuf, args: &HashMap<String, Value>) -> Result<ProvidedArgument> {
+    let path_arg = optional_arg!(
+        String,
+        args.get("path"),
+        GET_DATA_ARGUMENT_ERROR_MESSAGE
+    );
+
+    let url_arg = optional_arg!(
+        String,
+        args.get("url"),
+        GET_DATA_ARGUMENT_ERROR_MESSAGE
+    );
+
+    if !path_arg.is_some().bitxor(url_arg.is_some()) {
```

Comment:
I think it would be easier to read as `if path_arg.is_some() && url_arg.is_some()`
The case where both are empty will be taken care of at the end of the function